### PR TITLE
Малышев Владимир. TBB. Вариант 19. Поразрядная сортировка для вещественных чисел (тип double) с простым слиянием

### PR DESCRIPTION
**Pull request by dangerulya**
**Реализация поразрядной сортировки для вещественных чисел типа double с использованием TBB** 
**Класс SortTBB**  
Наследуется от `ppc::core::Task` и реализует этапы:  

**ValidationImpl**: Проверяет, что размер входного и выходного массивов совпадает.  
**PreProcessingImpl**: Копирует входные данные из `task_data` в буфер `input_`.  
**RunImpl**: Выполняет параллельную сортировку (основная логика).  
**PostProcessingImpl**: Копирует результат из `output_` в выходной массив `task_data`.  

---

**Функция ConvertDoubleToUInt64**  
Преобразует `double` в `uint64_t` для корректной работы с битовым представлением:  
- Если число отрицательное (старший бит = 1), инвертирует все биты.  
- Если положительное, устанавливает старший бит в 1.  
Это позволяет сортировать отрицательные числа в прямом порядке, а положительные — как беззнаковые.  

---

**Алгоритм сортировки**  

1. **Разделение чисел**:  
   - Входной массив разделяется на два вектора: `negative` (числа < 0) и `positive` (числа ≥ 0).  

2. **Параллельная обработка**:  
   - Каждый вектор обрабатывается в отдельном TBB-потоке с помощью `tbb::task_group`.  
   - Для сортировки используется функция `ProcessNumbers`, которая:  
     - Разбивает 64-битное представление числа на 8-битные группы (256 корзин).  
     - Для каждой группы (от младших к старшим битам):  
       - Подсчитывает элементы в корзинах.  
       - Перераспределяет элементы по корзинам.  

3. **Слияние результатов**:  
   - Отрицательные числа (уже отсортированные по возрастанию) объединяются с положительными (также по возрастанию).  

---

**Особенности реализации**

**Обработка отрицательных чисел**:  
   - Благодаря `ConvertDoubleToUInt64`, отрицательные числа сортируются как беззнаковые, но с сохранением их естественного порядка.  

 **Параллелизм с TBB**:  
   - Используется `tbb::task_arena` и `tbb::task_group` для параллельного выполнения сортировки `negative` и `positive`.  
   - Каждый блок (отрицательные/положительные) обрабатывается независимо, что увеличивает производительность на многопроцессорных системах.

### CODE REVIEW

**Review by mixcodi — CHANGES_REQUESTED** [review]

**Comment by dangerulya** [comment]
image

**Review by mixcodi — APPROVED** [review]

File: tasks/tbb/malyshev_v_radix_sort/src/ops_tbb.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+  oneapi::tbb::task_arena arena;
+  arena.execute([&] {
+    tbb::task_group tg;
+    tg.run([&] { ProcessNumbers(negative); });
+    tg.run([&] { ProcessNumbers(positive); });
+    tg.wait();
+  });
+
+  std::ranges::reverse(negative);
```
[the remark this answers is not in the document]
Comment (dangerulya) [reply tasks/tbb/malyshev_v_radix_sort/src/ops_tbb.cpp]: Done

File: tasks/tbb/malyshev_v_radix_sort/data/test.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+100
```
[the remark this answers is not in the document]
Comment (dangerulya) [reply tasks/tbb/malyshev_v_radix_sort/data/test.txt]: Done

File: tasks/tbb/malyshev_v_radix_sort/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+#include <algorithm>
+#include <cstdint>
+#include <cstdlib>
+#include <ctime>
+#include <memory>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+#include "tbb/malyshev_v_radix_sort/include/ops_tbb.hpp"
+
```
Comment (aobolensk) [inline tasks/tbb/malyshev_v_radix_sort/func_tests/main.cpp:12]: Please, add random functional tests as well

File: tasks/tbb/malyshev_v_radix_sort/perf_tests/main.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+#include "core/task/include/task.hpp"
+#include "tbb/malyshev_v_radix_sort/include/ops_tbb.hpp"
+
+TEST(malyshev_v_radix_sort_tbb, test_pipeline_run) {
+  constexpr int kCount = 1500000;
+  std::vector<double> in(kCount);
+  std::vector<double> out(kCount);
+
+  for (int i = 0; i < kCount; ++i) {
+    in[i] = static_cast<double>(rand()) / RAND_MAX * 1000.0;
```
Comment (aobolensk) [inline tasks/tbb/malyshev_v_radix_sort/perf_tests/main.cpp:20]: Please, also test negative numbers

File: tasks/tbb/malyshev_v_radix_sort/src/ops_tbb.cpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      positive.push_back(num);
+    }
+  }
+
+  oneapi::tbb::task_arena arena;
+  arena.execute([&] {
+    tbb::task_group tg;
+    tg.run([&] { ProcessNumbers(negative); });
+    tg.run([&] { ProcessNumbers(positive); });
+    tg.wait();
```
Comment (aobolensk) [inline tasks/tbb/malyshev_v_radix_sort/src/ops_tbb.cpp:89]: It seems you're using two threads only

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/tbb/malyshev_v_radix_sort/perf_tests/main.cpp (hunk 4 above)
Comment (dangerulya) [reply tasks/tbb/malyshev_v_radix_sort/perf_tests/main.cpp:20]: image

File: tasks/tbb/malyshev_v_radix_sort/func_tests/main.cpp (hunk 3 above)
Comment (dangerulya) [reply tasks/tbb/malyshev_v_radix_sort/func_tests/main.cpp:12]: image

Comment (aobolensk) [reply tasks/tbb/malyshev_v_radix_sort/func_tests/main.cpp:12]: I missed that, no need to add

File: tasks/tbb/malyshev_v_radix_sort/src/ops_tbb.cpp (hunk 5 above)
Comment (dangerulya) [reply tasks/tbb/malyshev_v_radix_sort/src/ops_tbb.cpp:105]: reworked

File: tasks/tbb/malyshev_v_radix_sort/perf_tests/main.cpp (hunk 4 above)
Comment (dangerulya) [reply tasks/tbb/malyshev_v_radix_sort/perf_tests/main.cpp]: done

**Review by aobolensk — APPROVED** [review]
